### PR TITLE
Dispatch accept/reject for rollover/collab settlement async for the maker

### PR DESCRIPTION
**Pull request by da-kami**
See commit message for details please.

Here are the logs that indicate (unless I'm mistaken) how often the maker_cfd actor run into this deadlock in production (the timeout then catches us, but we log an incorrect warning message...) in the last 48h:

```
2022-03-07 14:01:07	
{"timestamp":"2022-03-07T03:01:07.597476512Z","level":"WARN","fields":{"message":"Actor daemon::maker_cfd::Actor<daemon::oracle::Actor, daemon::maker_inc_connections::Actor, daemon::wallet::Actor> has been processing message daemon::maker_cfd::AcceptRollover for 10 seconds","log.target":"xtra::context","log.module_path":"xtra::context","log.file":"/home/runner/.cargo/git/checkouts/xtra-092fb79b35830946/95508e6/src/context.rs","log.line":314},"target":"xtra::context"}
	
2022-03-07 14:00:57	
{"timestamp":"2022-03-07T03:00:57.166536911Z","level":"WARN","fields":{"message":"Actor daemon::maker_cfd::Actor<daemon::oracle::Actor, daemon::maker_inc_connections::Actor, daemon::wallet::Actor> has been processing message daemon::maker_cfd::AcceptRollover for 10 seconds","log.target":"xtra::context","log.module_path":"xtra::context","log.file":"/home/runner/.cargo/git/checkouts/xtra-092fb79b35830946/95508e6/src/context.rs","log.line":314},"target":"xtra::context"}
	
2022-03-07 14:00:47	
{"timestamp":"2022-03-07T03:00:47.030317899Z","level":"WARN","fields":{"message":"Actor daemon::maker_cfd::Actor<daemon::oracle::Actor, daemon::maker_inc_connections::Actor, daemon::wallet::Actor> has been processing message daemon::maker_cfd::AcceptRollover for 10 seconds","log.target":"xtra::context","log.module_path":"xtra::context","log.file":"/home/runner/.cargo/git/checkouts/xtra-092fb79b35830946/95508e6/src/context.rs","log.line":314},"target":"xtra::context"}
	
2022-03-07 14:00:36	
{"timestamp":"2022-03-07T03:00:36.892364915Z","level":"WARN","fields":{"message":"Actor daemon::maker_cfd::Actor<daemon::oracle::Actor, daemon::maker_inc_connections::Actor, daemon::wallet::Actor> has been processing message daemon::maker_cfd::AcceptRollover for 10 seconds","log.target":"xtra::context","log.module_path":"xtra::context","log.file":"/home/runner/.cargo/git/checkouts/xtra-092fb79b35830946/95508e6/src/context.rs","log.line":314},"target":"xtra::context"}
	
2022-03-07 09:00:54	
{"timestamp":"2022-03-06T22:00:54.106867348Z","level":"WARN","fields":{"message":"Actor daemon::maker_cfd::Actor<daemon::oracle::Actor, daemon::maker_inc_connections::Actor, daemon::wallet::Actor> has been processing message daemon::maker_cfd::AcceptRollover for 10 seconds","log.target":"xtra::context","log.module_path":"xtra::context","log.file":"/home/runner/.cargo/git/checkouts/xtra-092fb79b35830946/95508e6/src/context.rs","log.line":314},"target":"xtra::context"}
	
2022-03-07 07:00:37	
{"timestamp":"2022-03-06T20:00:37.599654917Z","level":"WARN","fields":{"message":"Actor daemon::maker_cfd::Actor<daemon::oracle::Actor, daemon::maker_inc_connections::Actor, daemon::wallet::Actor> has been processing message daemon::maker_cfd::AcceptRollover for 10 seconds","log.target":"xtra::context","log.module_path":"xtra::context","log.file":"/home/runner/.cargo/git/checkouts/xtra-092fb79b35830946/95508e6/src/context.rs","log.line":314},"target":"xtra::context"}
	
2022-03-07 05:00:37	
{"timestamp":"2022-03-06T18:00:37.640292547Z","level":"WARN","fields":{"message":"Actor daemon::maker_cfd::Actor<daemon::oracle::Actor, daemon::maker_inc_connections::Actor, daemon::wallet::Actor> has been processing message daemon::maker_cfd::AcceptRollover for 10 seconds","log.target":"xtra::context","log.module_path":"xtra::context","log.file":"/home/runner/.cargo/git/checkouts/xtra-092fb79b35830946/95508e6/src/context.rs","log.line":314},"target":"xtra::context"}
	
2022-03-07 02:00:58	
{"timestamp":"2022-03-06T15:00:58.936723153Z","level":"WARN","fields":{"message":"Actor daemon::maker_cfd::Actor<daemon::oracle::Actor, daemon::maker_inc_connections::Actor, daemon::wallet::Actor> has been processing message daemon::maker_cfd::AcceptRollover for 10 seconds","log.target":"xtra::context","log.module_path":"xtra::context","log.file":"/home/runner/.cargo/git/checkouts/xtra-092fb79b35830946/95508e6/src/context.rs","log.line":314},"target":"xtra::context"}
	
2022-03-07 01:00:41	
{"timestamp":"2022-03-06T14:00:41.015538403Z","level":"WARN","fields":{"message":"Actor daemon::maker_cfd::Actor<daemon::oracle::Actor, daemon::maker_inc_connections::Actor, daemon::wallet::Actor> has been processing message daemon::maker_cfd::AcceptRollover for 10 seconds","log.target":"xtra::context","log.module_path":"xtra::context","log.file":"/home/runner/.cargo/git/checkouts/xtra-092fb79b35830946/95508e6/src/context.rs","log.line":314},"target":"xtra::context"}
	
2022-03-06 20:00:46	
{"timestamp":"2022-03-06T09:00:46.70325657Z","level":"WARN","fields":{"message":"Actor daemon::maker_cfd::Actor<daemon::oracle::Actor, daemon::maker_inc_connections::Actor, daemon::wallet::Actor> has been processing message daemon::maker_cfd::AcceptRollover for 10 seconds","log.target":"xtra::context","log.module_path":"xtra::context","log.file":"/home/runner/.cargo/git/checkouts/xtra-092fb79b35830946/95508e6/src/context.rs","log.line":314},"target":"xtra::context"}
	
2022-03-06 18:00:29	
{"timestamp":"2022-03-06T07:00:29.403915101Z","level":"WARN","fields":{"message":"Actor daemon::maker_cfd::Actor<daemon::oracle::Actor, daemon::maker_inc_connections::Actor, daemon::wallet::Actor> has been processing message daemon::maker_cfd::AcceptRollover for 10 seconds","log.target":"xtra::context","log.module_path":"xtra::context","log.file":"/home/runner/.cargo/git/checkouts/xtra-092fb79b35830946/95508e6/src/context.rs","log.line":314},"target":"xtra::context"}
	
2022-03-06 14:00:53	
{"timestamp":"2022-03-06T03:00:53.106310294Z","level":"WARN","fields":{"message":"Actor daemon::maker_cfd::Actor<daemon::oracle::Actor, daemon::maker_inc_connections::Actor, daemon::wallet::Actor> has been processing message daemon::maker_cfd::AcceptRollover for 10 seconds","log.target":"xtra::context","log.module_path":"xtra::context","log.file":"/home/runner/.cargo/git/checkouts/xtra-092fb79b35830946/95508e6/src/context.rs","log.line":314},"target":"xtra::context"}
	
2022-03-06 12:00:43	
{"timestamp":"2022-03-06T01:00:43.483378123Z","level":"WARN","fields":{"message":"Actor daemon::maker_cfd::Actor<daemon::oracle::Actor, daemon::maker_inc_connections::Actor, daemon::wallet::Actor> has been processing message daemon::maker_cfd::AcceptRollover for 10 seconds","log.target":"xtra::context","log.module_path":"xtra::context","log.file":"/home/runner/.cargo/git/checkouts/xtra-092fb79b35830946/95508e6/src/context.rs","log.line":314},"target":"xtra::context"}
	
2022-03-06 11:00:35	
{"timestamp":"2022-03-06T00:00:35.062113365Z","level":"WARN","fields":{"message":"Actor daemon::maker_cfd::Actor<daemon::oracle::Actor, daemon::maker_inc_connections::Actor, daemon::wallet::Actor> has been processing message daemon::maker_cfd::AcceptRollover for 10 seconds","log.target":"xtra::context","log.module_path":"xtra::context","log.file":"/home/runner/.cargo/git/checkouts/xtra-092fb79b35830946/95508e6/src/context.rs","log.line":314},"target":"xtra::context"}
	
2022-03-06 10:00:37	
{"timestamp":"2022-03-05T23:00:37.056146206Z","level":"WARN","fields":{"message":"Actor daemon::maker_cfd::Actor<daemon::oracle::Actor, daemon::maker_inc_connections::Actor, daemon::wallet::Actor> has been processing message daemon::maker_cfd::AcceptRollover for 10 seconds","log.target":"xtra::context","log.module_path":"xtra::context","log.file":"/home/runner/.cargo/git/checkouts/xtra-092fb79b35830946/95508e6/src/context.rs","log.line":314},"target":"xtra::context"}
	
2022-03-06 08:00:51	
{"timestamp":"2022-03-05T21:00:51.672299265Z","level":"WARN","fields":{"message":"Actor daemon::maker_cfd::Actor<daemon::oracle::Actor, daemon::maker_inc_connections::Actor, daemon::wallet::Actor> has been processing message daemon::maker_cfd::AcceptRollover for 10 seconds","log.target":"xtra::context","log.module_path":"xtra::context","log.file":"/home/runner/.cargo/git/checkouts/xtra-092fb79b35830946/95508e6/src/context.rs","log.line":314},"target":"xtra::context"}
	
2022-03-06 08:00:38	
{"timestamp":"2022-03-05T21:00:38.381183079Z","level":"WARN","fields":{"message":"Actor daemon::maker_cfd::Actor<daemon::oracle::Actor, daemon::maker_inc_connections::Actor, daemon::wallet::Actor> has been processing message daemon::maker_cfd::AcceptRollover for 10 seconds","log.target":"xtra::context","log.module_path":"xtra::context","log.file":"/home/runner/.cargo/git/checkouts/xtra-092fb79b35830946/95508e6/src/context.rs","log.line":314},"target":"xtra::context"}
	
2022-03-06 08:00:28	
{"timestamp":"2022-03-05T21:00:28.137284832Z","level":"WARN","fields":{"message":"Actor daemon::maker_cfd::Actor<daemon::oracle::Actor, daemon::maker_inc_connections::Actor, daemon::wallet::Actor> has been processing message daemon::maker_cfd::AcceptRollover for 10 seconds","log.target":"xtra::context","log.module_path":"xtra::context","log.file":"/home/runner/.cargo/git/checkouts/xtra-092fb79b35830946/95508e6/src/context.rs","log.line":314},"target":"xtra::context"}
	
2022-03-06 03:00:31	
{"timestamp":"2022-03-05T16:00:31.182973897Z","level":"WARN","fields":{"message":"Actor daemon::maker_cfd::Actor<daemon::oracle::Actor, daemon::maker_inc_connections::Actor, daemon::wallet::Actor> has been processing message daemon::maker_cfd::AcceptRollover for 10 seconds","log.target":"xtra::context","log.module_path":"xtra::context","log.file":"/home/runner/.cargo/git/checkouts/xtra-092fb79b35830946/95508e6/src/context.rs","log.line":314},"target":"xtra::context"}
	
2022-03-05 22:00:53	
{"timestamp":"2022-03-05T11:00:53.126873131Z","level":"WARN","fields":{"message":"Actor daemon::maker_cfd::Actor<daemon::oracle::Actor, daemon::maker_inc_connections::Actor, daemon::wallet::Actor> has been processing message daemon::maker_cfd::AcceptRollover for 10 seconds","log.target":"xtra::context","log.module_path":"xtra::context","log.file":"/home/runner/.cargo/git/checkouts/xtra-092fb79b35830946/95508e6/src/context.rs","log.line":314},"target":"xtra::context"}
	
2022-03-05 20:00:44	
{"timestamp":"2022-03-05T09:00:44.746041832Z","level":"WARN","fields":{"message":"Actor daemon::maker_cfd::Actor<daemon::oracle::Actor, daemon::maker_inc_connections::Actor, daemon::wallet::Actor> has been processing message daemon::maker_cfd::AcceptRollover for 10 seconds","log.target":"xtra::context","log.module_path":"xtra::context","log.file":"/home/runner/.cargo/git/checkouts/xtra-092fb79b35830946/95508e6/src/context.rs","log.line":314},"target":"xtra::context"}
	
2022-03-05 18:00:42	
{"timestamp":"2022-03-05T07:00:42.319238257Z","level":"WARN","fields":{"message":"Actor daemon::maker_cfd::Actor<daemon::oracle::Actor, daemon::maker_inc_connections::Actor, daemon::wallet::Actor> has been processing message daemon::maker_cfd::AcceptRollover for 10 seconds","log.target":"xtra::context","log.module_path":"xtra::context","log.file":"/home/runner/.cargo/git/checkouts/xtra-092fb79b35830946/95508e6/src/context.rs","log.line":314},"target":"xtra::context"}
```

### CODE REVIEW

**Comment by klochowicz** [comment]
> See commit message for details please.
> 
> Here are the logs that indicate (unless I'm mistaken) how often we run into the deadlock in production (the timeout then catches us, but we log an incorrect warning message...) in the last 48h:
> 
> ```
> 2022-03-07 14:01:07	
> {"timestamp":"2022-03-07T03:01:07.597476512Z","level":"WARN","fields":{"message":"Actor daemon::maker_cfd::Actor<daemon::oracle::Actor, daemon::maker_inc_connections::Actor, daemon::wallet::Actor> has been processing message daemon::maker_cfd::AcceptRollover for 10 seconds","log.target":"xtra::context","log.module_path":"xtra::context","log.file":"/home/runner/.cargo/git/checkouts/xtra-092fb79b35830946/95508e6/src/context.rs","log.line":314},"target":"xtra::context"}
> 	
> 2022-03-07 14:00:57	
> {"timestamp":"2022-03-07T03:00:57.166536911Z","level":"WARN","fields":{"message":"Actor daemon::maker_cfd::Actor<daemon::oracle::Actor, daemon::maker_inc_connections::Actor, daemon::wallet::Actor> has been processing message daemon::maker_cfd::AcceptRollover for 10 seconds","log.target":"xtra::context","log.module_path":"xtra::context","log.file":"/home/runner/.cargo/git/checkouts/xtra-092fb79b35830946/95508e6/src/context.rs","log.line":314},"target":"xtra::context"}
> 	
> 2022-03-07 14:00:47	
> {"timestamp":"2022-03-07T03:00:47.030317899Z","level":"WARN","fields":{"message":"Actor daemon::maker_cfd::Actor<daemon::oracle::Actor, daemon::maker_inc_connections::Actor, daemon::wallet::Actor> has been processing message daemon::maker_cfd::AcceptRollover for 10 seconds","log.target":"xtra::context","log.module_path":"xtra::context","log.file":"/home/runner/.cargo/git/checkouts/xtra-092fb79b35830946/95508e6/src/context.rs","log.line":314},"target":"xtra::context"}
> 	
> 2022-03-07 14:00:36	
> {"timestamp":"2022-03-07T03:00:36.892364915Z","level":"WARN","fields":{"message":"Actor daemon::maker_cfd::Actor<daemon::oracle::Actor, daemon::maker_inc_connections::Actor, daemon::wallet::Actor> has been processing message daemon::maker_cfd::AcceptRollover for 10 seconds","log.target":"xtra::context","log.module_path":"xtra::context","log.file":"/home/runner/.cargo/git/checkouts/xtra-092fb79b35830946/95508e6/src/context.rs","log.line":314},"target":"xtra::context"}
> 	
> 2022-03-07 09:00:54	
> {"timestamp":"2022-03-06T22:00:54.106867348Z","level":"WARN","fields":{"message":"Actor daemon::maker_cfd::Actor<daemon::oracle::Actor, daemon::maker_inc_connections::Actor, daemon::wallet::Actor> has been processing message daemon::maker_cfd::AcceptRollover for 10 seconds","log.target":"xtra::context","log.module_path":"xtra::context","log.file":"/home/runner/.cargo/git/checkouts/xtra-092fb79b35830946/95508e6/src/context.rs","log.line":314},"target":"xtra::context"}
> 	
> 2022-03-07 07:00:37	
> {"timestamp":"2022-03-06T20:00:37.599654917Z","level":"WARN","fields":{"message":"Actor daemon::maker_cfd::Actor<daemon::oracle::Actor, daemon::maker_inc_connections::Actor, daemon::wallet::Actor> has been processing message daemon::maker_cfd::AcceptRollover for 10 seconds","log.target":"xtra::context","log.module_path":"xtra::context","log.file":"/home/runner/.cargo/git/checkouts/xtra-092fb79b35830946/95508e6/src/context.rs","log.line":314},"target":"xtra::context"}
> 	
> 2022-03-07 05:00:37	
> {"timestamp":"2022-03-06T18:00:37.640292547Z","level":"WARN","fields":{"message":"Actor daemon::maker_cfd::Actor<daemon::oracle::Actor, daemon::maker_inc_connections::Actor, daemon::wallet::Actor> has been processing message daemon::maker_cfd::AcceptRollover for 10 seconds","log.target":"xtra::context","log.module_path":"xtra::context","log.file":"/home/runner/.cargo/git/checkouts/xtra-092fb79b35830946/95508e6/src/context.rs","log.line":314},"target":"xtra::context"}
> 	
> 2022-03-07 02:00:58	
> {"timestamp":"2022-03-06T15:00:58.936723153Z","level":"WARN","fields":{"message":"Actor daemon::maker_cfd::Actor<daemon::oracle::Actor, daemon::maker_inc_connections::Actor, daemon::wallet::Actor> has been processing message daemon::maker_cfd::AcceptRollover for 10 seconds","log.target":"xtra::context","log.module_path":"xtra::context","log.file":"/home/runner/.cargo/git/checkouts/xtra-092fb79b35830946/95508e6/src/context.rs","log.line":314},"target":"xtra::context"}
> 	
> 2022-03-07 01:00:41	
> {"timestamp":"2022-03-06T14:00:41.015538403Z","level":"WARN","fields":{"message":"Actor daemon::maker_cfd::Actor<daemon::oracle::Actor, daemon::maker_inc_connections::Actor, daemon::wallet::Actor> has been processing message daemon::maker_cfd::AcceptRollover for 10 seconds","log.target":"xtra::context","log.module_path":"xtra::context","log.file":"/home/runner/.cargo/git/checkouts/xtra-092fb79b35830946/95508e6/src/context.rs","log.line":314},"target":"xtra::context"}
> 	
> 2022-03-06 20:00:46	
> {"timestamp":"2022-03-06T09:00:46.70325657Z","level":"WARN","fields":{"message":"Actor daemon::maker_cfd::Actor<daemon::oracle::Actor, daemon::maker_inc_connections::Actor, daemon::wallet::Actor> has been processing message daemon::maker_cfd::AcceptRollover for 10 seconds","log.target":"xtra::context","log.module_path":"xtra::context","log.file":"/home/runner/.cargo/git/checkouts/xtra-092fb79b35830946/95508e6/src/context.rs","log.line":314},"target":"xtra::context"}
> 	
> 2022-03-06 18:00:29	
> {"timestamp":"2022-03-06T07:00:29.403915101Z","level":"WARN","fields":{"message":"Actor daemon::maker_cfd::Actor<daemon::oracle::Actor, daemon::maker_inc_connections::Actor, daemon::wallet::Actor> has been processing message daemon::maker_cfd::AcceptRollover for 10 seconds","log.target":"xtra::context","log.module_path":"xtra::context","log.file":"/home/runner/.cargo/git/checkouts/xtra-092fb79b35830946/95508e6/src/context.rs","log.line":314},"target":"xtra::context"}
> 	
> 2022-03-06 14:00:53	
> {"timestamp":"2022-03-06T03:00:53.106310294Z","level":"WARN","fields":{"message":"Actor daemon::maker_cfd::Actor<daemon::oracle::Actor, daemon::maker_inc_connections::Actor, daemon::wallet::Actor> has been processing message daemon::maker_cfd::AcceptRollover for 10 seconds","log.target":"xtra::context","log.module_path":"xtra::context","log.file":"/home/runner/.cargo/git/checkouts/xtra-092fb79b35830946/95508e6/src/context.rs","log.line":314},"target":"xtra::context"}
> 	
> 2022-03-06 12:00:43	
> {"timestamp":"2022-03-06T01:00:43.483378123Z","level":"WARN","fields":{"message":"Actor daemon::maker_cfd::Actor<daemon::oracle::Actor, daemon::maker_inc_connections::Actor, daemon::wallet::Actor> has been processing message daemon::maker_cfd::AcceptRollover for 10 seconds","log.target":"xtra::context","log.module_path":"xtra::context","log.file":"/home/runner/.cargo/git/checkouts/xtra-092fb79b35830946/95508e6/src/context.rs","log.line":314},"target":"xtra::context"}
> 	
> 2022-03-06 11:00:35	
> {"timestamp":"2022-03-06T00:00:35.062113365Z","level":"WARN","fields":{"message":"Actor daemon::maker_cfd::Actor<daemon::oracle::Actor, daemon::maker_inc_connections::Actor, daemon::wallet::Actor> has been processing message daemon::maker_cfd::AcceptRollover for 10 seconds","log.target":"xtra::context","log.module_path":"xtra::context","log.file":"/home/runner/.cargo/git/checkouts/xtra-092fb79b35830946/95508e6/src/context.rs","log.line":314},"target":"xtra::context"}
> 	
> 2022-03-06 10:00:37	
> {"timestamp":"2022-03-05T23:00:37.056146206Z","level":"WARN","fields":{"message":"Actor daemon::maker_cfd::Actor<daemon::oracle::Actor, daemon::maker_inc_connections::Actor, daemon::wallet::Actor> has been processing message daemon::maker_cfd::AcceptRollover for 10 seconds","log.target":"xtra::context","log.module_path":"xtra::context","log.file":"/home/runner/.cargo/git/checkouts/xtra-092fb79b35830946/95508e6/src/context.rs","log.line":314},"target":"xtra::context"}
> 	
> 2022-03-06 08:00:51	
> {"timestamp":"2022-03-05T21:00:51.672299265Z","level":"WARN","fields":{"message":"Actor daemon::maker_cfd::Actor<daemon::oracle::Actor, daemon::maker_inc_connections::Actor, daemon::wallet::Actor> has been processing message daemon::maker_cfd::AcceptRollover for 10 seconds","log.target":"xtra::context","log.module_path":"xtra::context","log.file":"/home/runner/.cargo/git/checkouts/xtra-092fb79b35830946/95508e6/src/context.rs","log.line":314},"target":"xtra::context"}
> 	
> 2022-03-06 08:00:38	
> {"timestamp":"2022-03-05T21:00:38.381183079Z","level":"WARN","fields":{"message":"Actor daemon::maker_cfd::Actor<daemon::oracle::Actor, daemon::maker_inc_connections::Actor, daemon::wallet::Actor> has been processing message daemon::maker_cfd::AcceptRollover for 10 seconds","log.target":"xtra::context","log.module_path":"xtra::context","log.file":"/home/runner/.cargo/git/checkouts/xtra-092fb79b35830946/95508e6/src/context.rs","log.line":314},"target":"xtra::context"}
> 	
> 2022-03-06 08:00:28	
> {"timestamp":"2022-03-05T21:00:28.137284832Z","level":"WARN","fields":{"message":"Actor daemon::maker_cfd::Actor<daemon::oracle::Actor, daemon::maker_inc_connections::Actor, daemon::wallet::Actor> has been processing message daemon::maker_cfd::AcceptRollover for 10 seconds","log.target":"xtra::context","log.module_path":"xtra::context","log.file":"/home/runner/.cargo/git/checkouts/xtra-092fb79b35830946/95508e6/src/context.rs","log.line":314},"target":"xtra::context"}
> 	
> 2022-03-06 03:00:31	
> {"timestamp":"2022-03-05T16:00:31.182973897Z","level":"WARN","fields":{"message":"Actor daemon::maker_cfd::Actor<daemon::oracle::Actor, daemon::maker_inc_connections::Actor, daemon::wallet::Actor> has been processing message daemon::maker_cfd::AcceptRollover for 10 seconds","log.target":"xtra::context","log.module_path":"xtra::context","log.file":"/home/runner/.cargo/git/checkouts/xtra-092fb79b35830946/95508e6/src/context.rs","log.line":314},"target":"xtra::context"}
> 	
> 2022-03-05 22:00:53	
> {"timestamp":"2022-03-05T11:00:53.126873131Z","level":"WARN","fields":{"message":"Actor daemon::maker_cfd::Actor<daemon::oracle::Actor, daemon::maker_inc_connections::Actor, daemon::wallet::Actor> has been processing message daemon::maker_cfd::AcceptRollover for 10 seconds","log.target":"xtra::context","log.module_path":"xtra::context","log.file":"/home/runner/.cargo/git/checkouts/xtra-092fb79b35830946/95508e6/src/context.rs","log.line":314},"target":"xtra::context"}
> 	
> 2022-03-05 20:00:44	
> {"timestamp":"2022-03-05T09:00:44.746041832Z","level":"WARN","fields":{"message":"Actor daemon::maker_cfd::Actor<daemon::oracle::Actor, daemon::maker_inc_connections::Actor, daemon::wallet::Actor> has been processing message daemon::maker_cfd::AcceptRollover for 10 seconds","log.target":"xtra::context","log.module_path":"xtra::context","log.file":"/home/runner/.cargo/git/checkouts/xtra-092fb79b35830946/95508e6/src/context.rs","log.line":314},"target":"xtra::context"}
> 	
> 2022-03-05 18:00:42	
> {"timestamp":"2022-03-05T07:00:42.319238257Z","level":"WARN","fields":{"message":"Actor daemon::maker_cfd::Actor<daemon::oracle::Actor, daemon::maker_inc_connections::Actor, daemon::wallet::Actor> has been processing message daemon::maker_cfd::AcceptRollover for 10 seconds","log.target":"xtra::context","log.module_path":"xtra::context","log.file":"/home/runner/.cargo/git/checkouts/xtra-092fb79b35830946/95508e6/src/context.rs","log.line":314},"target":"xtra::context"}
> ```

but are we in the deadlock or does it wait until it finishes processing previous message? it seems that it never gets stuck for longer?

**Comment by da-kami** [comment]
>> snip
> 
> but are we in the deadlock or does it wait until it finishes processing previous message? it seems that it never gets stuck for longer?

This is not a critical fix in the sense of "nothing works, we have to fix this", because the timeout catches us. But blocking the `maker_cfd` actor for `10` seconds on several occasions is still bad. We cannot process settlements, rollovers, new offers, (...) when blocked.

Seeing these messages so often is an indicator that there is in fact a deadlock and we should just dispatch async. That's why I created this PR.

**Comment by klochowicz** [comment]
> > > snip
> > 
> > 
> > but are we in the deadlock or does it wait until it finishes processing previous message? it seems that it never gets stuck for longer?
> 
> This is not a critical fix in the sense of "nothing works, we have to fix this", because the timeout catches us. But blocking the `maker_cfd` actor for `10` seconds on several occasions is still bad. We cannot process settlements, rollovers, new offers, (...) when blocked.
> 
> Seeing these messages so often is an indicator that there is in fact a deadlock and we should just dispatch async. That's why I created this PR.

I agree that blocking for 10s is a bug that we should fix, I was just opposed to calling it a "deadlock", which, by definition, is a situation from which you can't get out of.  :)

**Comment by da-kami** [comment]
> > > > snip
> > > 
> > > 
> > > but are we in the deadlock or does it wait until it finishes processing previous message? it seems that it never gets stuck for longer?
> > 
> > 
> > This is not a critical fix in the sense of "nothing works, we have to fix this", because the timeout catches us. But blocking the `maker_cfd` actor for `10` seconds on several occasions is still bad. We cannot process settlements, rollovers, new offers, (...) when blocked.
> > Seeing these messages so often is an indicator that there is in fact a deadlock and we should just dispatch async. That's why I created this PR.
> 
> I agree that blocking for 10s is a bug that we should fix, I was just opposed to calling it a "deadlock", which, by definition, is a situation from which you can't get out of. :)

I think the actor still runs into a deadlock, the application is fine because we added the timeouts. Can try to describe that in more detail next time. I updated the description to mention that the deadlock is on the actor.

**Comment by klochowicz** [comment]
> > > > > snip
> > > > 
> > > > 
> > > > but are we in the deadlock or does it wait until it finishes processing previous message? it seems that it never gets stuck for longer?
> > > 
> > > 
> > > This is not a critical fix in the sense of "nothing works, we have to fix this", because the timeout catches us. But blocking the `maker_cfd` actor for `10` seconds on several occasions is still bad. We cannot process settlements, rollovers, new offers, (...) when blocked.
> > > Seeing these messages so often is an indicator that there is in fact a deadlock and we should just dispatch async. That's why I created this PR.
> > 
> > 
> > I agree that blocking for 10s is a bug that we should fix, I was just opposed to calling it a "deadlock", which, by definition, is a situation from which you can't get out of. :)
> 
> I think the actor still runs into a deadlock, the application is fine because we added the timeouts. Can try to describe that in more detail next time. I updated the description to mention that the deadlock is on the actor.

again, where do you see the deadlock? maker_cfd actor does not get restarted, so it should never process any other messages after the deadlock. You can't have dozens of "deadlocks" and things picking up eventually on their own.
To me the situation above is more of a bottleneck, where things get queued up during processing, which blocks some tasks temporarily.

**Comment by thomaseizinger** [comment]
There is merit in both your viewpoints:

1. Deadlocks are typically unrecoverable.
2. We might run into a deadlock that eventually clears up due to the timeouts we added to the futures. I am not sure if one would still call it a deadlock then?
3. Unless we see an error for a timeout, it does not seem like we run into an actual deadlock but just slow processing of messages.

In regards to this code change: Sending messages async can solve deadlocks but it cannot solve slow processing of messages. If we actually queue up several messages to be processed by an actor, the offending message that created the initial slow down will still not process faster.

**Comment by da-kami** [comment]
> 
> In regards to this code change: Sending messages async can solve deadlocks but it cannot solve slow processing of messages. If we actually queue up several messages to be processed by an actor, the offending message that created the initial slow down will still not process faster.

I don't think slow processing is what happens here. I think this is literally a case where we were "stuck forever" before we added the timeouts. I think this code change is a fix, not just a change.

**Comment by klochowicz** [comment]
> > In regards to this code change: Sending messages async can solve deadlocks but it cannot solve slow processing of messages. If we actually queue up several messages to be processed by an actor, the offending message that created the initial slow down will still not process faster.
> 
> I don't think slow processing is what happens here. I think this is literally a case where we were "stuck forever" before we added the timeouts. I think this code change is a fix, not just a change.

OK, let's re-cap this. I think you tried to say "we would have run into a deadlock if we didn't have the timeouts", which might be the case. I was just arguing that we are not running into a deadlock, because this would stall our whole application.

File: daemon/src/maker_cfd.rs [hunk 1]
```diff
@@ -394,41 +395,61 @@ impl<O, T, W> Actor<O, T, W> {
...
+            Ok(Ok(())) => Ok(()),
+            Ok(Err(NotConnected(e))) => {
+                self.executor
+                    .execute(order_id, |cfd| Ok(cfd.fail_contract_setup(anyhow!(e))))
+                    .await?;
 
-            bail!("Accept failed: No active contract setup for order {order_id}")
+                bail!("Accept failed: No active contract setup for order {order_id}")
+            }
+            Err(ellapsed) => {
```
Comment (klochowicz) [inline daemon/src/maker_cfd.rs]: nit: spelling - "elapsed"

File: daemon/src/maker_cfd.rs [hunk 2]
```diff
@@ -394,41 +395,61 @@ impl<O, T, W> Actor<O, T, W> {
...
+            Ok(Ok(())) => Ok(()),
+            Ok(Err(NotConnected(e))) => {
+                self.executor
+                    .execute(order_id, |cfd| Ok(cfd.fail_contract_setup(anyhow!(e))))
+                    .await?;
 
-            bail!("Reject failed: No active contract setup for order {order_id}")
+                bail!("Reject failed: No active contract setup for order {order_id}")
+            }
+            Err(ellapsed) => {
```
Comment (klochowicz) [inline daemon/src/maker_cfd.rs]: nit: "elapsed"

File: daemon/src/maker_cfd.rs [hunk 3]
```diff
@@ -44,10 +44,6 @@ use xtras::SendAsyncSafe;
```
Comment (klochowicz) [inline daemon/src/maker_cfd.rs:44]: re commit description: I'd suggest softer phrasing than "we definitely have a deadlock" when we are do not have 100% clarity due to incomplete logs

File: daemon/src/maker_cfd.rs (hunk 1 above)
Comment (thomaseizinger) [inline daemon/src/maker_cfd.rs]: Nit: Typo in `Elapsed`.

**Review by thomaseizinger — APPROVED** [review]

**Review by klochowicz — APPROVED** [review]
nice job! 
shall we consider backporting this on top of `0.4.7` ?

**Pull request merged**